include/RAJA/policy/cuda/atomic.hpp

```diff
@@ -43,596 +51,549 @@ namespace RAJA
+  T expected;
 
+  do {
+    expected = old;
+    old = cuda_atomicCAS(acc, expected, oper(expected));
+  } while (!cuda_atomicCAS_equal(old, expected));
 
-// 32-bit unsigned atomicMax support by CUDA
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicMax<unsigned>(unsigned volatile *acc,
-                                                    unsigned value)
-{
-  return ::atomicMax((unsigned *)acc, value);
+  return old;
 }
-#endif
 
-// 64-bit unsigned atomicMax support by CUDA sm_35 and later
-#if __CUDA_ARCH__ >= 350
-template <>
-RAJA_INLINE __device__ unsigned long long cuda_atomicMax<unsigned long long>(
-    unsigned long long volatile *acc,
-    unsigned long long value)
+/*!
+ * Generic impementation of any atomic 32-bit or 64-bit operator with short-circuiting.
+ * Implementation uses the existing CUDA supplied unsigned 32-bit or 64-bit CAS
+ * operator. Returns the OLD value that was replaced by the result of this
+ * operation.
+ */
+template <typename T, typename Oper, typename ShortCircuit>
+RAJA_INLINE __device__ T cuda_atomicCAS(T *acc, Oper&& oper, ShortCircuit&& sc)
 {
-  return ::atomicMax((unsigned long long *)acc, value);
-}
-#endif
+  T old = cuda_atomicLoad(acc);
 
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicInc(T volatile *acc, T val)
-{
-  // See:
-  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicinc
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T old) {
-    return ((old >= val) ? 0 : (old + 1));
-  });
-}
+  if (sc(old)) {
+    return old;
+  }
 
-// 32-bit unsigned atomicInc support by CUDA
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicInc<unsigned>(unsigned volatile *acc,
-                                                    unsigned value)
-{
-  return ::atomicInc((unsigned *)acc, value);
-}
+  T expected;
 
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicInc(T volatile *acc)
-{
-  return cuda_atomic_CAS_oper(acc,
-                                      [=] __device__(T a) { return a + 1; });
-}
+  do {
+    expected = old;
+    old = cuda_atomicCAS(acc, expected, oper(expected));
+  } while (!cuda_atomicCAS_equal(old, expected) && !sc(old));
 
-// 32-bit signed atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ int cuda_atomicInc<int>(int volatile *acc)
-{
-  return ::atomicAdd((int *)acc, (int)1);
+  return old;
 }
 
-// 32-bit unsigned atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicInc<unsigned>(unsigned volatile *acc)
-{
-  return ::atomicAdd((unsigned *)acc, (unsigned)1);
-}
 
-// 64-bit unsigned atomicAdd support by CUDA, used as backend for atomicInc
-template <>
-RAJA_INLINE __device__ unsigned long long cuda_atomicInc<unsigned long long>(
-    unsigned long long volatile *acc)
+/*!
+ * Atomic addition
+ */
+using cuda_atomicAdd_builtin_types = list<
+      int
+     ,unsigned int
```

Comment:
Is it better to use a "leading comma" *only* for the macro guarded places and use trailing comma in all other places? For example, I would probably use trailing commas here and then do the following below.....

Reply:
Fixed.